### PR TITLE
added shopping cart and shopping cart items

### PR DESCRIPTION
**Pull request by DmytroFedoryshyn**

### CODE REVIEW

**Review by Rommelua — CHANGES_REQUESTED** [review]
You have 112 files in your PR including **target** folder
It should be in gitignore
Remove all extra files from PR
image

File: bookstore.iml
Comment (Rommelua) [inline bookstore.iml:1]: do we need this file?

File: src/main/java/bookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+@RequestMapping("/api/cart")
+@RequiredArgsConstructor
+@Tag(name = "Shopping Cart Controller", description = "Endpoints for managing the shopping cart")
+public class ShoppingCartController {
+    private final ShoppingCartService service;
+
+    @Operation(summary = "Get the user's shopping cart")
+    @GetMapping
+    public CartResponseDto get() {
+        return service.getByUser();
```
Comment (Rommelua) [inline src/main/java/bookstore/controller/ShoppingCartController.java]: Fetch User on Controller layer (**for all methods**)
```suggestion
        User user = (User) authentication.getPrincipal();
        return service.getById(user.getId());
```

File: src/main/java/bookstore/dto/cartitem/UpdateCartItemDto.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package bookstore.dto.cartitem;
+
+import lombok.Data;
+
+@Data
+public class UpdateCartItemDto {
```
Comment (Rommelua) [inline src/main/java/bookstore/dto/cartitem/UpdateCartItemDto.java]: same as AddToCartDto 
Do we need two classes?

File: src/main/java/bookstore/repository/cartitem/CartItemRepository.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package bookstore.repository.cartitem;
+
+import bookstore.model.CartItem;
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.repository.query.Param;
+
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
+    @EntityGraph(attributePaths = {"book"})
+    CartItem getReferenceById(@Param("cartItemId") Long cartItemId);
```
Comment (Rommelua) [inline src/main/java/bookstore/repository/cartitem/CartItemRepository.java]: Reference? Why not CartItem?

File: src/main/java/bookstore/service/ShoppingCartServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        ShoppingCart cart = getCartByUser();
+        return cartMapper.toDto(cart);
+    }
+
+    @Override
+    public void addItemToCart(AddToCartDto dto) {
+        CartItem cartItem = cartMapper.toCartItem(dto);
+        ShoppingCart cart = getCartByUser();
+        cart.getItems().add(cartItem);
+        cartItem.setShoppingCart(cart);
```
Comment (Rommelua) [inline src/main/java/bookstore/service/ShoppingCartServiceImpl.java]: why do we need setShoppingCart?

File: src/main/java/bookstore/service/ShoppingCartService.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import bookstore.dto.cartitem.UpdateCartItemDto;
+
+public interface ShoppingCartService {
+    CartResponseDto getByUser();
+
+    void addItemToCart(AddToCartDto dto);
+
+    CartItemResponseDto update(Long id, UpdateCartItemDto dto);
+
+    void delete(Long id);
```
Comment (Rommelua) [inline src/main/java/bookstore/service/ShoppingCartService.java]: delete ShoppingCart?

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/bookstore/controller/ShoppingCartController.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+@Tag(name = "Shopping Cart Controller", description = "Endpoints for managing the shopping cart")
+public class ShoppingCartController {
+    private final ShoppingCartService service;
+
+    @Operation(summary = "Get the user's shopping cart")
+    @GetMapping
+    public CartResponseDto get() {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        return service.getShoppingCartByUser((User) authentication.getPrincipal());
+    }
```
Comment (sokimaaa) [inline src/main/java/bookstore/controller/ShoppingCartController.java]:
```suggestion
    public CartResponseDto get(Authentication authentication) {
        User user = (User) authentication.getPrincipal();
        return service.getShoppingCartByUser(user);
    }
```

File: src/main/java/bookstore/controller/ShoppingCartController.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            @Schema(implementation = CartItemResponseDto.class),
+                    examples = @ExampleObject(value = "{\"id\": 1,"
+                            + " \"book\": {\"id\": 1, \"title\":"
+                            + " \"Book Title\"}, \"quantity\": 2}"))
+    })
+    @ApiResponse(responseCode = "404", description = "Cart item not found")
+    public CartItemResponseDto updateCartItem(
+            @PathVariable @Parameter(description = "ID of the cart item to update") Long id,
+            @RequestBody @Parameter(description = "Updated cart item details")
+                    AddToCartDto dto) {
```
Comment (sokimaaa) [inline src/main/java/bookstore/controller/ShoppingCartController.java]:
```suggestion
            @RequestBody @Parameter(description = "Updated cart item details")
                  @Valid  AddToCartDto dto) {
```

Let's add validation for this dto, e.g. quantity > 0.

File: src/main/java/bookstore/controller/ShoppingCartController.java (hunk 6 above)
Comment (sokimaaa) [reply src/main/java/bookstore/controller/ShoppingCartController.java]: the same for other endpoints

File: src/main/java/bookstore/controller/ShoppingCartController.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    @PostMapping()
+    @ApiResponse(responseCode = "204", description = "Item added to the cart")
+    public void addCartItemTo(Authentication authentication,
+            @RequestBody @Parameter(description = "Item to add to the cart")
+            @Valid AddToCartDto addToCartDto) {
+        service.addItemToCart(addToCartDto, (User) authentication.getPrincipal());
+    }
+
+    @Operation(summary = "Update a shopping cart item by ID")
+    @PutMapping("/cart-items/{id}")
```
Comment (Rommelua) [inline src/main/java/bookstore/controller/ShoppingCartController.java:53]: image

File: src/main/java/bookstore/controller/ShoppingCartController.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+                    examples = @ExampleObject(value = "{\"id\": 1,"
+                            + " \"book\": {\"id\": 1, \"title\":"
+                            + " \"Book Title\"}, \"quantity\": 2}"))
+    })
+    @ApiResponse(responseCode = "404", description = "Cart item not found")
+    public CartItemResponseDto updateCartItem(
+            @PathVariable @Parameter(description = "ID of the cart item to update") Long id,
+            @RequestBody @Parameter(description = "Updated cart item details")
+            @Valid AddToCartDto dto) {
+        return service.updateCartItem(id, dto);
```
Comment (Rommelua) [inline src/main/java/bookstore/controller/ShoppingCartController.java:73]: let's add more security - we should check if this cart item belongs to this users' Cart
same for delete

File: src/main/java/bookstore/model/ShoppingCart.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(nullable = false)
+    private Long id;
+
+    @OneToOne
+    private User user;
+
+    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
+    private final Set<CartItem> items = new HashSet<>();
```
Comment (Rommelua) [inline src/main/java/bookstore/model/ShoppingCart.java:37]: add equals to CartItem class

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/bookstore/controller/ShoppingCartController.java (hunk 8 above)
Comment (DmytroFedoryshyn) [reply src/main/java/bookstore/controller/ShoppingCartController.java:53]: I have "api/cart" part in my @RequestMapping annotation placed on top of controller

File: src/main/java/bookstore/model/CartItem.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    private Book book;
+
+    @Column(nullable = false)
+    private int quantity;
+
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+
+    @Override
+    public boolean equals(Object obj) {
```
Comment (Rommelua) [inline src/main/java/bookstore/model/CartItem.java:41]: we must override hashCode() as well

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/bookstore/model/CartItem.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        CartItem other = (CartItem) obj;
+
+        return id.equals(other.id);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, book, quantity);
```
Comment (Rommelua) [inline src/main/java/bookstore/model/CartItem.java:57]:
```suggestion
        return Objects.hash(id);
```

**Pull request merged**